DATAJDBC-256 - Run integration tests with Oracle.

schauder opened the pull request:
An Oracle database is started in Docker using Testcontainers.
Tests that don't work yet are skipped using the TestDatabaseFeatures which allows central control, which database supports which feature.

A suitable oracle image is deployed on DockerHub as  "springci/spring-data-oracle-xe-prebuild:18.4.0".
 The official Oracle docker images as described here https://github.com/oracle/docker-images/blob/master/OracleDatabase/SingleInstance/README.md are not suitable since it needs about 15min to start and also TestContainers seems to break it by trying to us it to early.
 The referenced image was created based on these instructions: https://github.com/oracle/docker-images/tree/master/OracleDatabase/SingleInstance/samples/prebuiltdb

The following features don't yet work for Oracle:

* Loading of date like properties.
* Ids with custom names.
* Array support.
* Boolean properties.
* BigDecimals and BigInteger is limited to what fits into Oracles NUMBER data type.
* Entities that use a join during load time, i.e. with a one to one relationship.


See DATAJDBC-256.

mp911de commented on spring-data-relational/src/main/java/org/springframework/data/relational/core/dialect/IdGeneration.java:
```diff
@@ -0,0 +1,35 @@
+	 * @return {@literal true} if the a list of column names should get passed to the JDBC driver for which ids shall be
+	 *         generated.
+	 */
+	default boolean setKeyColumnNames() {
```
How about `driverRequiresKeyColumnNames` with a reference to `@see Connection#prepareStatement(String, String[])`?

schauder replied:
done

mp911de commented on spring-data-jdbc-distribution/pom.xml:
```diff
@@ -14,7 +14,7 @@
 		<groupId>org.springframework.data</groupId>
 		<artifactId>spring-data-relational-parent</artifactId>
-		<version>2.1.0-SNAPSHOT</version>
+		<version>2.1.0.DATAJDBC-256</version>
```
That should be `2.1.0-DATAJDBC-256-SNAPSHOT` to avoid that these artifacts don't get deployed into the release/milestone repository.

schauder replied:
😬. Fixed.

mp911de commented on pom.xml:
```diff
@@ -190,6 +190,24 @@
 									</systemPropertyVariables>
 								</configuration>
 							</execution>
+							<execution>
```
Container startup fails with a TNS error message quite regularly. That causes a lot of the CI builds to fail.

I'm wondering whether we should run DB2 and Oracle tests only in a nightly job.

schauder replied:
Where do you see these regular failures? In Jenkins I see one failure in the last 14 builds which would gives me hope that just waiting a little longer might fix the problem, especially since it is a TNS failure and not a "your account has been locked".

schauder commented:
> We could make use of `JdbcUtils.getResultSetValue(…)` and the actual property type to delegate result value conversion to the driver

Sounds very promising.  I didn't know that method! But I'd leave that to DATAJDBC-569 and limit this PR to the just getting the very basics working as agreed before if that is ok.

schauder commented on spring-data-relational/src/main/java/org/springframework/data/relational/core/dialect/IdGeneration.java:
```diff
@@ -0,0 +1,35 @@
+ */
+package org.springframework.data.relational.core.dialect;
+
+public interface IdGeneration {
```
done.

schauder commented on spring-data-jdbc/src/test/java/org/springframework/data/jdbc/testing/OracleDataSourceConfiguration.java:
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
```
Done.

schauder commented on spring-data-jdbc/src/test/java/org/springframework/data/jdbc/core/JdbcAggregateTemplateIntegrationTests.java:
```diff
@@ -199,6 +188,8 @@ private static LegoSet createLegoSet(String name) {
 	@Test // DATAJDBC-112
 	public void saveAndLoadAnEntityWithReferencedEntityById() {
 
+		features.supportsQuotedIds();
```
Interesting. I have the very opposite impression. The direct call to `features` is shorter both in the underlying implementation and on the use case site and easier to understand.

I'll prepare a commit switching to a rule based approach.

steveswing commented on spring-data-relational/src/main/java/org/springframework/data/relational/core/dialect/OracleDialect.java:
```diff
@@ -0,0 +1,55 @@
+import org.springframework.util.ClassUtils;
+
+/**
+ * An SQL dialect for Postgres.
```
Should be Oracle?

schauder replied:
Good find. Thanks.

schauder commented:
@mp911de 
I addressed the various comments.

I left the migration to annotation based assumptions in a separate commit.
It actually doesn't look as bad as I expected :-)

I also rebased on the current master and removed the _alias for join with as_ since that is already resolved.

For the CI build stability I increased the maximum wait time to 5minutes for the Oracle db to start up.
Let's see if that helps or if we have to come up with something else.

mp911de commented:
That's merged and polished now.
